### DISCUSSION TITLE
OpenXR - Invalid Layer Error

### DISCUSSION THREAD

**Discussion by courgeon** [dotnet/Silk.NET#1097] [Q&A] [answered]
Hello everyone  !

I'm back to trying to work with Silk.NET, OpenXR (Windows + Vive Cosmos XR runtime) and OpenGL

I've managed to get quite far now, I initialize everything, get the headset position and views properties and so on.
But i'm now stuck in my attempt to provide images to the headset.

the process is : 

```
WaitFrame  
LocateView
SyncAction

BeginFrame
   for each eye
       AcquireSwapchainImage (color)
       WaitSwapchainImage (color)
       AcquireSwapchainImage (depth)
       WaitSwapchainImage (depth)
       Bind the FBO with provided Image ID and depth ID
       // render stuff
       ReleaseSwapchainImage(color)
       ReleaseSwapchainImage(depth)
```
up to this, every single xr call returns with Success.

then, i try to finalize frame and send it the the headset, The code is inspired by the OpenGL XR  demo of Silk.NET

```
                  var frameEndInfo = new FrameEndInfo(
                      displayTime: framestate.PredictedDisplayTime,
                      environmentBlendMode: EnvironmentBlendMode.Opaque
                  );

                var projectionLayerViews = stackalloc CompositionLayerProjectionView[2];
                var projectionLayer = new CompositionLayerProjection
                (
                    type: StructureType.CompositionLayerProjection,
                    viewCount: 2,
                    views: projectionLayerViews,
                    space: play_space
                );


                for (var eye = 0; eye < 2; eye++)
                {
                    projectionLayerViews[eye] = new CompositionLayerProjectionView();
                    projectionLayerViews[eye].SubImage.Swapchain = swapchains[eye];
                    projectionLayerViews[eye].SubImage.ImageRect.Extent = new
                    (
                        (int)viewconfig_views[eye].RecommendedImageRectWidth,
                        (int)viewconfig_views[eye].RecommendedImageRectHeight
                    );

                    projectionLayerViews[eye].SubImage.ImageRect.Offset.X = 0;
                    projectionLayerViews[eye].SubImage.ImageRect.Offset.Y = 0;
                    projectionLayerViews[eye].Fov = projection_views[eye].Fov;
                    projectionLayerViews[eye].Pose = projection_views[eye].Pose;
                }

                var layerPointer = (CompositionLayerBaseHeader*)&projectionLayer;

                frameEndInfo.LayerCount = 1;
                frameEndInfo.Layers = &layerPointer;

                result = Xr.EndFrame(session, in frameEndInfo);
                CheckFrame(result, "failed to end frame : " + result.ToString());
```

But EndFrame fails with "ErrorLayerInvalid"

I really don't know how to progress from here.

The OpenXR specs says 

> XR_ERROR_LAYER_INVALID must be returned if an unknown, unsupported layer type, or NULL pointer is passed as one of the layers.
> XR_ERROR_LAYER_INVALID must be returned if a layer references a swapchain that has no released swapchain image.

But as far as i can tell, it's seems that it's not the case for either of these condition.

Any idea would really help !

Thanks

**Comment by Perksey** [comment]
What example are you using? We don't have any OpenXR examples.

**Reply by courgeon** [reply]
It's not on the master branch, but I found this

https://github.com/dotnet/Silk.NET/blob/b0b31779ce4db9b68922977fa11772b95f506e09/examples/CSharp/OpenGL%20Demos/OpenGL%20VR%20Demo/OpenXR/Renderer.cs#L507

It was linked in a comments of this project

https://github.com/phr00t/FocusEngine/blob/master/sources/engine/Xenko.VirtualReality/OpenXR/OpenXRHmd.cs

**Comment by courgeon** [comment, marked as answer]
Ok... I figured out my mistake. 
Because of some copy paste error, I was using xrImages generated by the left-eye swapchain in the right-eye swapchain. *Dumb me*

Pro tip : Just don't ! :) 

Too bad the OpenXR API is not more specific when it comes to signaling errors.
Anyway, it works now with Oculus XR runtime (Quest 2+link), Steam XR runtime (Vive cosmos), but I get a Memory violation on xrEndFrame using the Vive XR Runtime. 
It's probably not related to Silk though, I suspect that the runtime is at fault here.

M.

**Reply by Perksey** [reply]
Nice, that's excellent news!
